src/storage-client/src/storage_collections.rs

```diff
@@ -2037,25 +2354,20 @@ where
                             }
 
                         }
-                        BackgroundCmd::DowngradeSince(cmds) => {
-                            self.downgrade_sinces(cmds).await;
-                        }
-                        BackgroundCmd::Update { id, is_in_txns, write_handle, since_handle } => {
-                            self.shard_by_id.insert(id, write_handle.shard_id()).expect(
-                                "BackgroundCmd::Update only valid for updating extant write handles",
-                            );
+                        BackgroundCmd::Upgrade { id, since_handle } => {
+                            assert!(matches!(since_handle, SinceHandleWrapper::Critical(_)), "can only upgrade a leased handle to a critical handle");
 
-                            self.since_handles.insert(id, since_handle).expect(
-                                "BackgroundCmd::Update only valid for updating extant since handles",
-                            );
+                            let previous = self.since_handles.insert(id, since_handle);
 
-                            if is_in_txns {
-                                self.txns_shards.insert(id);
+                            if let Some(previous) = previous {
+                                assert!(matches!(previous, SinceHandleWrapper::Leased(_)), "can only upgrade a leased handle to a critical handle");
+                                previous.expire().await;
                             } else {
-                                let fut = gen_upper_future(id, write_handle);
-                                upper_futures.push(fut.boxed());
+                                 panic!("BackgroundCmd::Upgrade only valid for updating extant since handles");
```

Comment:
Nit: The indentation is off here.

Reply:
stupid rustfmt not working inside `select!` blocks ... 🤷‍♂️